File: packages/core-ts/src/gate.ts
/**
 * The pre-read gate.
 *
 * Runs before the underlying tool executes, so a blocked decision means the host tool was
 * never invoked and no oversized payload ever existed. The decision is tri-state:
 *
 * - `passthrough` - the host owns the call; the gate could not prove a safe intervention
 * - `allow`       - read-like and provably within a configured bounded form
 * - `blocked`     - an unbounded read of a regular file is provably over the full-read cap
 *
 * Sizing comes from a bounded probe that stops at 351 lines or the byte cap and carries
 * its own 1s deadline. A probe can only block a full read when its observed lower bound
 * proves that the source is over a configured cap; uncertainty passes through to the host.
 */
import { Clock, Deadline, monotonicClock } from "./clock.js";
import { DEFAULT_LIMITS, Limits } from "./limits.js";
import { Classification, classifyCommand } from "./shell.js";

export type Decision = "passthrough" | "allow" | "blocked";
export type GateForm =
  | "not_read_like"
  | "full_read"
  | "bounded_lines"
  | "bounded_search"
  | "bounded_metadata"
  | "unclassifiable"
  | "unsafe";

export interface ProbeResult {
  readonly exists: boolean;
  readonly kind?: "file" | "directory" | "fifo" | "socket" | "device" | "other";
  readonly lines?: number;
  readonly bytes?: number;
  readonly exact?: boolean;
}

export type ProbeSelection =
  | { readonly mode: "full" }
  | { readonly mode: "metadata" }
  | { readonly mode: "lines"; readonly startLine: number; readonly limit: number }
  | { readonly mode: "tail"; readonly limit: number }
  | { readonly mode: "search"; readonly maxMatches: number };

export type Prober = (path: string, selection?: ProbeSelection) => ProbeResult;

export interface GateDecision {
  readonly decision: Decision;
  readonly form: GateForm;
  readonly code?: string;
  readonly reason: string;
  readonly observedLines?: number;
  readonly observedBytes?: number;
}

export interface ToolArgs {
  [key: string]: unknown;
}

const PASSTHROUGH: GateDecision = { decision: "passthrough", form: "not_read_like", reason: "" };

function blocked(code: string, form: GateForm, reason: string, extra: Partial<GateDecision> = {}): GateDecision {
  return { decision: "blocked", form, code, reason, ...extra };
}

function passthrough(
  form: GateForm = "not_read_like",
  reason = "",
  extra: Partial<GateDecision> = {},
): GateDecision {
  return { decision: "passthrough", form, reason, ...extra };
}

function allow(form: GateForm): GateDecision {
  return { decision: "allow", form, reason: "" };
}

interface SizedProbe {
  lines: number;
  bytes: number;
  exact: boolean;
}

export class PreReadGate {
  constructor(
    private readonly probe: Prober,
    private readonly limits: Limits = DEFAULT_LIMITS,
    private readonly clock: Clock = monotonicClock,
  ) {}

  evaluate(tool: string, args: ToolArgs): GateDecision {
    try {
      const deadline = Deadline.start(this.clock, this.limits.gateProbeDeadlineMs);
      if (tool === "read") return this.evaluateRead(args, deadline);
      if (tool === "search") return this.evaluateSearch(args, deadline);
      if (tool === "shell") return this.evaluateShell(String(args["command"] ?? ""), deadline);
      return PASSTHROUGH;
    } catch (err) {
      // An unfinished probe does not establish a large source. Fail open and let the
      // host report its own bounded tool result or error.
      if (err instanceof Error && (err as { code?: string }).code === "TIMEOUT") {
        return passthrough("unclassifiable", "PROBE_TIMEOUT");
      }
      if (err instanceof Error && (err as { code?: string }).code === "CANCELLED") {
        return passthrough("unclassifiable", "PROBE_CANCELLED");
      }
      // The gate is an advisory context-cost control. A probe or malformed adapter
      // input must never turn an otherwise valid host call into a synthetic error.
      return passthrough("unclassifiable", "PROBE_FAILED");
    }
  }

  private evaluateRead(args: ToolArgs, deadline: Deadline): GateDecision {
    const path = args["file_path"] ?? args["path"];
    if (typeof path !== "string" || path.length === 0) return passthrough("unclassifiable", "BAD_PATH");

    const limit = args["limit"];
    const offset = args["offset"];
    for (const [name, value] of [["LIMIT", limit], ["OFFSET", offset]] as const) {
      if (value !== undefined && value !== null && (typeof value !== "number" || !Number.isInteger(value))) {
        return passthrough("unclassifiable", `BAD_${name}`);
      }
    }
    if (typeof limit === "number" && limit < 1) {
      return passthrough("unclassifiable", "BAD_LIMIT");
    }
    if (typeof offset === "number" && offset < 0) {
      return passthrough("unclassifiable", "BAD_OFFSET");
    }

    // Any explicit limit is a bounded host call. A limit above our preferred page
    // size is still finite; the gate must not reinterpret it as an unbounded read.
    if (typeof limit === "number" && limit > this.limits.targetedReadMaxLines) {
      return passthrough("bounded_lines", "BOUND_OVER_CAP");
    }

    const bounded = typeof limit === "number";
    if (bounded) {
      const start = typeof offset === "number" ? Math.max(1, offset) : 1;
      // Cover both 0- and 1-based host offset conventions. The extra line is used only
      // for byte proof; the host is still capped at `limit` output lines.
      const scanLimit = limit + (typeof offset === "number" && offset > 0 ? 1 : 0);
      const selected = this.probeSafe(path, deadline, {
        mode: "lines",
        startLine: start,
        limit: scanLimit,
      });
      if (selected === null) return PASSTHROUGH;
      if ("decision" in selected) return selected;
      return this.sizeBounded([selected], "bounded_lines", limit);
    }
    const probed = this.probeSafe(path, deadline);
    if (probed === null) return PASSTHROUGH;
    if ("decision" in probed) return probed;
    return this.sizeFullRead([probed]);
  }

  private evaluateSearch(args: ToolArgs, deadline: Deadline): GateDecision {
    if (
      (args["target"] ?? "content") !== "content"
      || (args["output_mode"] ?? "content") !== "content"
      || (args["context"] ?? 0) !== 0
    ) {
      return passthrough("unclassifiable", "OUTPUT_AMPLIFICATION");
    }
    const maxMatches = args["max_matches"];
    if (
      typeof maxMatches !== "number" ||
      !Number.isInteger(maxMatches) ||
      maxMatches < 1 ||
      maxMatches > this.limits.targetedSearchMaxMatches
    ) {
      return passthrough("unclassifiable", "UNBOUNDED_SEARCH");
    }
    const pattern = args["pattern"];
    if (typeof pattern !== "string" || pattern.length === 0) {
      return passthrough("unclassifiable", "BAD_PATTERN");
    }
    const path = args["path"] ?? args["file_path"];
    if (typeof path !== "string" || path.length === 0) {
      return passthrough("unclassifiable", "UNRESOLVED_PATH");
    }
    const probed = this.probeSafe(path, deadline, { mode: "search", maxMatches });
    if (probed === null) return PASSTHROUGH;
    if ("decision" in probed) return probed;
    return this.sizeBounded([probed], "bounded_search", maxMatches);
  }

  private evaluateShell(command: string, deadline: Deadline): GateDecision {
    const c: Classification = classifyCommand(command);
    if (c.form === "not_read_like") return PASSTHROUGH;
    if (c.form === "unclassifiable") {
      return passthrough("unclassifiable", c.reason || "UNPROVABLE");
    }

    const probes: SizedProbe[] = [];
    let selection: ProbeSelection = { mode: "full" };
    let boundedLines = 0;
    if (c.form === "bounded_metadata") selection = { mode: "metadata" };
    if (c.form === "bounded_search") {
      if ((c.boundMatches ?? 0) > this.limits.targetedSearchMaxMatches) {
        return passthrough("bounded_search", "BOUND_OVER_CAP");
      }
      selection = { mode: "search", maxMatches: c.boundMatches ?? 0 };
    }
    if (c.form === "bounded_lines") {
      const bound = c.boundLines ?? 0;
      boundedLines = bound * c.files.length + (c.files.length > 1 ? c.files.length * 2 : 0);
      if (bound < 1 || boundedLines > this.limits.targetedReadMaxLines) {
        return passthrough("bounded_lines", "BOUND_OVER_CAP", {
          observedLines: boundedLines,
        });
      }
      selection = c.fromEnd
        ? { mode: "tail", limit: bound }
        : { mode: "lines", startLine: c.lineStart ?? 1, limit: bound };
    }
    if (c.form === "bounded_metadata") {
      const fields = c.metadataFields ?? 3;
      let projectedBytes = c.files.length > 1 ? 64 : 0;
      for (const path of c.files) {
        projectedBytes += new TextEncoder().encode(path).length + fields * 24 + 16;
        if (
          c.files.length > this.limits.targetedReadMaxLines
          || projectedBytes > this.limits.maxTargetedReadBytes
        ) {
          return passthrough("bounded_metadata", "OUTPUT_OVER_CAP", {
            observedLines: c.files.length,
            observedBytes: projectedBytes,
          });
        }
      }
    }
    for (const path of c.files) {
      const probed = this.probeSafe(path, deadline, selection);
      if (probed === null) continue;
      if ("decision" in probed) return probed;
      probes.push(probed);
    }
    if (probes.length === 0) return PASSTHROUGH;

    if (c.form === "bounded_metadata") return allow("bounded_metadata");
    if (c.form === "bounded_search") {
      return this.sizeBounded(probes, "bounded_search", c.boundMatches ?? 0);
    }
    if (c.form === "bounded_lines") {
      const headerBytes = c.files.length > 1
        ? c.files.reduce((total, path) => total + new TextEncoder().encode(path).length + 16, 0)
        : 0;
      return this.sizeBounded(probes, "bounded_lines", boundedLines, headerBytes);
    }
    return this.sizeFullRead(probes);
  }

  /** `null` means "let the host handle it" (missing file). */
  private probeSafe(
    path: string,
    deadline: Deadline,
    selection: ProbeSelection = { mode: "full" },
  ): SizedProbe | GateDecision | null {
    deadline.check("PROBE");
    const probe = this.probe(path, selection);
    // Checked again after the scan: an over-budget probe cannot justify a veto.
    deadline.check("PROBE");
    if (!probe.exists) return null;
    if ((probe.kind ?? "file") !== "file") {
      return passthrough("unsafe", "NOT_REGULAR_FILE");
    }
    const lines = probe.lines ?? 0;
    const bytes = probe.bytes ?? 0;
    const exact = probe.exact ?? true;
    if (
      !Number.isSafeInteger(lines) || lines < 0
      || !Number.isSafeInteger(bytes) || bytes < 0
      || typeof exact !== "boolean"
    ) {
      return passthrough("unclassifiable", "INVALID_PROBE");
    }
    return { lines, bytes, exact };
  }

  private sizeBounded(
    probes: SizedProbe[],
    form: "bounded_lines" | "bounded_search",
    outputLines: number,
    extraBytes = 0,
  ): GateDecision {
    const totalBytes = probes.reduce((sum, p) => sum + p.bytes, extraBytes);
    if (probes.some((p) => !p.exact)) {
      return passthrough(form, "UNKNOWN_SCALE", { observedLines: outputLines });
    }
    if (totalBytes > this.limits.maxTargetedReadBytes) {
      return passthrough(form, "OVER_BYTE_THRESHOLD", {
        observedLines: outputLines,
        observedBytes: totalBytes,
      });
    }
    return allow(form);
  }

  private sizeFullRead(probes: SizedProbe[]): GateDecision {
    const totalLines = probes.reduce((sum, p) => sum + p.lines, 0);
    const totalBytes = probes.reduce((sum, p) => sum + p.bytes, 0);
    const inexact = probes.some((p) => !p.exact);
    if (totalBytes > this.limits.maxTargetedReadBytes) {
      return blocked("LARGE_READ", "full_read", "OVER_BYTE_THRESHOLD", {
        observedLines: totalLines,
        observedBytes: totalBytes,
      });
    }
    if (inexact || totalLines > this.limits.fullReadMaxLines) {
      if (totalLines > this.limits.fullReadMaxLines) {
        return blocked("LARGE_READ", "full_read", "OVER_LINE_THRESHOLD", {
          observedLines: totalLines,
          observedBytes: totalBytes,
        });
      }
      // An inexact result with no lower-bound evidence is uncertainty, not a veto.
      return passthrough("full_read", "UNKNOWN_SCALE", {
        observedLines: totalLines,
        observedBytes: totalBytes,
      });
    }
    return allow("full_read");
  }
}

export const GUIDANCE_LARGE_READ =
  "This source is over the full-read threshold, so the read was stopped before it ran. " +
  "Re-read a specific range with offset+limit (<=350 lines), run a bounded search, or ask " +
  "the context-shunt reader a question about it - the reader answers from the source " +
  "without putting it in this conversation.";

export const GUIDANCE_UNCLASSIFIABLE_READ =
  "This command reads a file but could not be proven bounded and safe, so it was stopped " +
  "before it ran. Use a plain read with offset+limit, a bounded search, or ask the " +
  "context-shunt reader a question about the file.";

export const GUIDANCE_UNSAFE_SOURCE =
  "The target is not a regular file, so it was not read. Name a regular file inside a " +
  "configured workspace root.";

export function guidanceFor(decision: GateDecision): string {
  if (decision.code === "LARGE_READ") return GUIDANCE_LARGE_READ;
  if (decision.code === "UNCLASSIFIABLE_READ") return GUIDANCE_UNCLASSIFIABLE_READ;
  return GUIDANCE_UNSAFE_SOURCE;
}

File: packages/core-py/src/context_shunt/gate.py
"""The pre-read gate.

Runs *before* the underlying tool executes, so a blocked decision means the host tool
was never invoked and no oversized payload ever existed. The decision is tri-state:

``passthrough``  the call is not read-like (or the file does not exist); host policy owns it
``allow``        read-like and provably small or provably bounded
``blocked``      ``LARGE_READ`` / ``UNCLASSIFIABLE_READ`` / ``UNSAFE_SOURCE``

Sizing comes from a bounded probe that stops at 351 lines or the byte cap and carries
its own 1s deadline; an inexact probe is "unknown scale", which blocks.
"""

from __future__ import annotations

from dataclasses import dataclass
from enum import Enum
from typing import Any, Protocol

from .clock import Clock, Deadline, MonotonicClock
from .errors import DeadlineExceeded
from .limits import DEFAULT_LIMITS, Limits
from .shell import Classification, ShellForm, classify_command


class Decision(str, Enum):
    PASSTHROUGH = "passthrough"
    ALLOW = "allow"
    BLOCKED = "blocked"


class GateForm(str, Enum):
    NOT_READ_LIKE = "not_read_like"
    FULL_READ = "full_read"
    BOUNDED_LINES = "bounded_lines"
    BOUNDED_SEARCH = "bounded_search"
    BOUNDED_METADATA = "bounded_metadata"
    UNCLASSIFIABLE = "unclassifiable"
    UNSAFE = "unsafe"


@dataclass(frozen=True)
class ProbeResult:
    exists: bool
    kind: str = "file"  # file | directory | fifo | socket | device | other
    lines: int = 0
    bytes: int = 0
    exact: bool = True


class Prober(Protocol):
    def __call__(self, path: str) -> ProbeResult: ...


@dataclass(frozen=True)
class GateDecision:
    decision: Decision
    form: GateForm
    code: str | None = None
    reason: str = ""
    # Non-sensitive sizing for the blocked guidance. Never the file content.
    observed_lines: int | None = None
    observed_bytes: int | None = None

    @property
    def blocked(self) -> bool:
        return self.decision is Decision.BLOCKED


_ALLOW_NOT_READ = GateDecision(Decision.PASSTHROUGH, GateForm.NOT_READ_LIKE)


def _blocked(code: str, form: GateForm, reason: str, **kw: Any) -> GateDecision:
    return GateDecision(Decision.BLOCKED, form, code=code, reason=reason, **kw)


class PreReadGate:
    def __init__(
        self,
        prober: Prober,
        limits: Limits = DEFAULT_LIMITS,
        clock: Clock | None = None,
    ):
        self._probe = prober
        self._limits = limits
        self._clock = clock or MonotonicClock()

    # -- public ------------------------------------------------------------
    def evaluate(self, tool: str, args: dict[str, Any]) -> GateDecision:
        deadline = Deadline.start(self._clock, self._limits.gate_probe_deadline_ms)
        try:
            if tool == "read":
                return self._evaluate_read(args, deadline)
            if tool == "search":
                return self._evaluate_search(args)
            if tool == "shell":
                return self._evaluate_shell(str(args.get("command") or ""), deadline)
            return _ALLOW_NOT_READ
        except DeadlineExceeded:
            # An unfinished probe means unknown scale, which blocks rather than executes.
            return _blocked("LARGE_READ", GateForm.FULL_READ, "PROBE_TIMEOUT")

    # -- read tool ---------------------------------------------------------
    def _evaluate_read(self, args: dict[str, Any], deadline: Deadline) -> GateDecision:
        path = args.get("file_path") or args.get("path")
        if not path or not isinstance(path, str):
            return _ALLOW_NOT_READ
        limit = args.get("limit")
        offset = args.get("offset")
        for name, value in (("limit", limit), ("offset", offset)):
            if value is not None and (not isinstance(value, int) or isinstance(value, bool)):
                return _blocked("UNCLASSIFIABLE_READ", GateForm.UNCLASSIFIABLE, f"BAD_{name.upper()}")
        if limit is not None and limit < 1:
            return _blocked("UNCLASSIFIABLE_READ", GateForm.UNCLASSIFIABLE, "BAD_LIMIT")
        if offset is not None and offset < 0:
            return _blocked("UNCLASSIFIABLE_READ", GateForm.UNCLASSIFIABLE, "BAD_OFFSET")

        bounded = limit is not None and limit <= self._limits.targeted_read_max_lines
        probe = self._probe_safe(path, deadline)
        if probe is None:
            return _ALLOW_NOT_READ
        if isinstance(probe, GateDecision):
            return probe
        if bounded:
            return GateDecision(Decision.ALLOW, GateForm.BOUNDED_LINES)
        return self._size_full_read([probe])

    # -- search tool -------------------------------------------------------
    def _evaluate_search(self, args: dict[str, Any]) -> GateDecision:
        max_matches = args.get("max_matches")
        if (
            max_matches is None
            or isinstance(max_matches, bool)
            or not isinstance(max_matches, int)
            or max_matches < 1
            or max_matches > self._limits.targeted_search_max_matches
        ):
            return _blocked("UNCLASSIFIABLE_READ", GateForm.UNCLASSIFIABLE, "UNBOUNDED_SEARCH")
        pattern = args.get("pattern")
        if not isinstance(pattern, str) or not pattern:
            return _blocked("UNCLASSIFIABLE_READ", GateForm.UNCLASSIFIABLE, "BAD_PATTERN")
        return GateDecision(Decision.ALLOW, GateForm.BOUNDED_SEARCH)

    # -- shell -------------------------------------------------------------
    def _evaluate_shell(self, command: str, deadline: Deadline) -> GateDecision:
        c: Classification = classify_command(command)
        if c.form is ShellForm.NOT_READ_LIKE:
            return _ALLOW_NOT_READ
        if c.form is ShellForm.UNCLASSIFIABLE:
            return _blocked("UNCLASSIFIABLE_READ", GateForm.UNCLASSIFIABLE, c.reason or "UNPROVABLE")

        probes: list[ProbeResult] = []
        for path in c.files:
            probe = self._probe_safe(path, deadline)
            if probe is None:
                return _ALLOW_NOT_READ
            if isinstance(probe, GateDecision):
                return probe
            probes.append(probe)

        if c.form is ShellForm.BOUNDED_METADATA:
            return GateDecision(Decision.ALLOW, GateForm.BOUNDED_METADATA)
        if c.form is ShellForm.BOUNDED_SEARCH:
            if (c.bound_matches or 0) > self._limits.targeted_search_max_matches:
                return _blocked("UNCLASSIFIABLE_READ", GateForm.UNCLASSIFIABLE, "UNBOUNDED_SEARCH")
            return GateDecision(Decision.ALLOW, GateForm.BOUNDED_SEARCH)
        if c.form is ShellForm.BOUNDED_LINES:
            bound = c.bound_lines or 0
            if bound > self._limits.targeted_read_max_lines:
                return _blocked(
                    "LARGE_READ", GateForm.BOUNDED_LINES, "BOUND_OVER_CAP", observed_lines=bound
                )
            return GateDecision(Decision.ALLOW, GateForm.BOUNDED_LINES)
        return self._size_full_read(probes)

    # -- helpers -----------------------------------------------------------
    def _probe_safe(self, path: str, deadline: Deadline) -> ProbeResult | GateDecision | None:
        """Probe one path. ``None`` means "let the host handle it" (missing file)."""
        deadline.check("PROBE")
        probe = self._probe(path)
        # Checked again after the scan: a probe that overran the budget leaves the
        # overall scale unknown, and unknown scale blocks.
        deadline.check("PROBE")
        if not probe.exists:
            return None
        if probe.kind != "file":
            return _blocked("UNSAFE_SOURCE", GateForm.UNSAFE, "NOT_REGULAR_FILE")
        return probe

    def _size_full_read(self, probes: list[ProbeResult]) -> GateDecision:
        total_lines = sum(p.lines for p in probes)
        total_bytes = sum(p.bytes for p in probes)
        inexact = any(not p.exact for p in probes)
        if inexact or total_lines > self._limits.full_read_max_lines:
            return _blocked(
                "LARGE_READ",
                GateForm.FULL_READ,
                "OVER_LINE_THRESHOLD" if not inexact else "UNKNOWN_SCALE",
                observed_lines=None if inexact else total_lines,
                observed_bytes=None if inexact else total_bytes,
            )
        if total_bytes > self._limits.max_targeted_read_bytes:
            return _blocked(
                "LARGE_READ",
                GateForm.FULL_READ,
                "OVER_BYTE_THRESHOLD",
                observed_lines=total_lines,
                observed_bytes=total_bytes,
            )
        return GateDecision(Decision.ALLOW, GateForm.FULL_READ)


GUIDANCE_LARGE_READ = (
    "This source is over the full-read threshold, so the read was stopped before it ran. "
    "Re-read a specific range with offset+limit (<=350 lines), run a bounded search, or ask "
    "the context-shunt reader a question about it - the reader answers from the source "
    "without putting it in this conversation."
)
GUIDANCE_UNCLASSIFIABLE_READ = (
    "This command reads a file but could not be proven bounded and safe, so it was stopped "
    "before it ran. Use a plain read with offset+limit, a bounded search, or ask the "
    "context-shunt reader a question about the file."
)
GUIDANCE_UNSAFE_SOURCE = (
    "The target is not a regular file, so it was not read. Name a regular file inside a "
    "configured workspace root."
)


def guidance_for(decision: GateDecision) -> str:
    if decision.code == "LARGE_READ":
        return GUIDANCE_LARGE_READ
    if decision.code == "UNCLASSIFIABLE_READ":
        return GUIDANCE_UNCLASSIFIABLE_READ
    return GUIDANCE_UNSAFE_SOURCE
